Addition of OutgoingFrameMaxSize Ice2 connection parameter

bentoi opened the pull request:
This PR adds support for the Ice2 OutgoingFrameMaxSize connection parameter. The value of this parameter is checked when sending a request or response to ensure it's not larger than the peer's incoming frame max size. If it's larger, a new remote MarshalException is raised.

bernardnormier commented on slice/Ice/Exceptions.ice:
```diff
@@ -63,4 +63,10 @@ module Ice
         string paramName;
     }
+
+    /// A marshal exception is thrown when a request or response can't be encoded. This can occur if the
```
When a request cannot be written, Ice should throw a local exception (that is, a System.Exception that is not a ZeroC.Ice.RemoteException). Not being able to write a request is purely local. When your invocation fails with a remote exception, the expectation is the request was sent. (Therefore, when LocatorInfo fails to resolve an endpoint, it should not throw or let through a remote exception, but that's a different topic).

So here the remote MarshalException should be used only when Ice fails to marshal a response or exception? For the future Data frames from server to client, I assume a marshaling failure just resets the stream and does not send an exception.

I am also wondering if MarshalException is the best exception name here.

We already have UnhandledException for unexpected errors such as:
 - failing to unmarshal request parameters
 - the user code throws an exception that is not a RemoteException
 - the user code throws an remote exception with ConvertToUnhandled set to true

UnhandledException represents a bug in the application, so if we consider that when the marshaling of a response or exception exceeds the peer's IncomingFrameMaxSize is not a bug but "expected", it's desirable to use a different remote exception. Maybe `LimitExceededException` - an exception thrown by the remote peer when some limit is exceeded? Or `ImplementationException` - an exception thrown by the remote peer's implementation that is not unhandled, so may not be a bug?

bentoi replied:
Before you suggested to use MarshalException, the implementation originally raised ArgumentOutOfRange exception if the request couldn't be written when too large. What local exception do you suggest to raise here instead? To me ArgumentOutOfRangeException or ArgumentException with an appropriate message still works since the exception is the result of arguments being too large to encode.

I don't have a strong preference for the remote exception... I don't think UnhandledException is necessarily bad either if we want to keep it simple. To me the line is quite thin on whether or not we should consider this as an application bug or not. So what do you prefer?

bernardnormier commented on csharp/src/Ice/MultiStreamSocket.cs:
```diff
@@ -32,6 +32,7 @@ public abstract class MultiStreamSocket : IDisposable
         public bool IsIncoming { get; }
 
         internal int IncomingFrameMaxSize { get; }
+        internal int? OutgoingFrameMaxSize { get; set; }
```
Would be good to add a doc-comment describing these properties.

My understanding in the OutgoingFrameMaxSize is the peer's IncomingFrameMaxSize and therefore is only set during connection establishment.

bernardnormier replied:
We could also rename OutgoingFrameMaxSize => PeerIncomingFrameMaxSize?

bernardnormier commented on slice/Ice/Ice2Definitions.ice:
```diff
@@ -28,6 +28,12 @@ module Ice
+    /// The keys for supported Ice2 connection parameters.
+    unchecked enum Ice2ParameterKey : int
+    {
+        OutgoingFrameMaxSize = 0
```
I think it would be more logical to call this parameter `IncomingFrameMaxSize`: the sender provides its IncomingFrameMaxSize, and the recipient interprets this size as its own OutgoingFrameMaxSize.

bernardnormier commented on csharp/src/Ice/SocketStream.cs:
```diff
@@ -300,9 +315,12 @@ internal virtual async ValueTask SendInitializeFrameAsync(CancellationToken canc
 
-                // TODO: send protocol specific parameters from the frame?
+                // The outgoing frame maximum size for the peer is the local incoming frame maximum size
+                ostr.WriteBinaryContextEntry((int)Ice2ParameterKey.OutgoingFrameMaxSize,
```
See comment on Ice2ParameterKey.

bernardnormier commented on csharp/src/Ice/SocketStream.cs:
```diff
@@ -383,6 +401,20 @@ internal virtual async ValueTask SendInitializeFrameAsync(CancellationToken canc
+            {
+                if (frame is OutgoingRequestFrame)
+                {
+                    throw new MarshalException($"the request size is larger than {_socket.OutgoingFrameMaxSize}");
```
I think this should be a local exception. Then, when forwarding requests (e.g. ForwardAsync), the forwarder can convert some local exceptions into remote exceptions. See comment on MarshalException.

bernardnormier commented on csharp/src/Ice/SocketStream.cs:
```diff
@@ -383,6 +401,20 @@ internal virtual async ValueTask SendInitializeFrameAsync(CancellationToken canc
+                {
+                    // Throw a remote exception instead for response, the Ice connection will catch it and send it
+                    // as the response instead of sending this response which is too large.
+                    throw new MarshalException($"the response size is larger than {_socket.OutgoingFrameMaxSize}");
```
For the error message will be something like `the response size is larger than 1048576`?

I am concerned it won't be clear to the reader where this number of coming from. `the response size is larger than the caller's IncomingFrameMaxSize (1048576 bytes)`?

bentoi replied:
Fixed to `$@"the request size is larger than the peer's IncomingFrameSizeMax ({_socket.PeerIncomingFrameMaxSize} bytes)'`

bernardnormier commented on csharp/src/Ice/SlicSocket.cs:
```diff
@@ -23,8 +23,9 @@ public override TimeSpan IdleTimeout
 
         internal int BidirectionalStreamCount;
         internal AsyncSemaphore? BidirectionalStreamSemaphore;
+        internal int MaxBidirectionalStreams;
```
In general for non private fields / properties, we prefer properties.

bentoi replied:
I'll fix this in another PR.

bernardnormier commented on csharp/src/Ice/SlicSocket.cs:
```diff
@@ -23,8 +23,9 @@ public override TimeSpan IdleTimeout
         internal AsyncSemaphore? BidirectionalStreamSemaphore;
+        internal int MaxBidirectionalStreams;
+        internal int MaxUnidirectionalStreams;
         internal readonly object Mutex = new();
```
Not directly related to this PR: it's desirable for mutexes and other synchronization objects to be private, not shared between several classes.

bentoi replied:
I'll fix this in another PR.

bernardnormier commented on csharp/test/Ice/exceptions/AllTests.cs:
```diff
@@ -262,25 +262,42 @@ public static IThrowerPrx Run(TestHelper helper)
                     thrower.ThrowMemoryLimitException(Array.Empty<byte>());
```
Once we settle on a name for this exception, we should rename the operation.

externl commented on slice/Ice/Ice2Definitions.ice:
```diff
@@ -28,6 +28,12 @@ module Ice
+    /// The keys for supported Ice2 connection parameters.
+    unchecked enum Ice2ParameterKey : int
+    {
+        OutgoingFrameMaxSize = 0
```
It's not obvious from the name direction of the connection this applies to. Does it apply to both client and server? Is there any negotiation between the client and server's OutgoingFrameMaxSize?

bentoi replied:
It's renamed IncomngFrameMaxSize now.

externl commented on slice/Ice/Exceptions.ice:
```diff
@@ -63,4 +63,10 @@ module Ice
+
+    /// A marshal exception is thrown when a request or response can't be encoded. This can occur if the
+    /// request or response exceeds the frame maximum size for example.
+    exception MarshalException
+    {
+    }
```
I don't see exceeding the frame limit size as the response can't be encoded. Maybe `FrameSizeLimitException` is better.

bentoi replied:
I used LimitExceededException for now.

bernardnormier commented on csharp/src/Ice/Communicator.cs:
```diff
@@ -502,25 +504,24 @@ private class ObserverUpdater : Instrumentation.IObserverUpdater
+                if (SlicPacketMaxSize < 1024)
                 {
-                    throw new InvalidConfigurationException($"{SlicOptions.PacketMaxSize} is not a valid " +
+                    throw new InvalidConfigurationException($"{SlicPacketMaxSize} is not a valid " +
```
should probably not use `+` either.

bentoi commented on csharp/src/Ice/Communicator.cs:
```diff
@@ -502,25 +504,24 @@ private class ObserverUpdater : Instrumentation.IObserverUpdater
-                if (SlicOptions.PacketMaxSize < 1024)
+
+                SlicPacketMaxSize = GetPropertyAsInt("Ice.Slic.PacketMaxSize") ?? 32 * 1024;
+                if (SlicPacketMaxSize < 1024)
```
Specifying something smaller doesn't seem very useful to me... Quic uses 1200 for its packet size to limit fragmentation  of the UDP datagram. Slic is based on TCP and the idea of doing the fragmentation at the Slic level with packets of 1KB could be to avoid TCP fragmentation... however, I don't know if this is really useful either as I didn't really do any performance testing or investigate in which use cases this could be beneficial. Do you perhaps have another suggestion for this limit and the default? I think we should eventually run some performance testing for the default value.

externl replied:
`1024` seems good to me!

bentoi commented on csharp/src/Ice/SocketStream.cs:
```diff
@@ -180,8 +180,23 @@ internal virtual async ValueTask ReceiveInitializeFrameAsync(CancellationToken c
+                    {
+                        checked
+                        {
+                            _socket.OutgoingFrameMaxSize = (int)value.Span.ReadVarULong().Value;
```
Good point. We don't impose a minimum size for `Ice.IncomingFrameMaxSize` though. The old `Ice.MessageSizeMax` minimum was 1KB. Should we add back this 1KB minimum limit?

externl replied:
Yeah, I guess `1KB` makes the most sense as that's the smallest max you can specify through configuration right?

bentoi replied:
We did accept values < 1KB before. I changed it to reject them with this PR.

bentoi commented:
Pushed fixes:
- renamed OutgoingFrameMaxSize to PeerIncomingFrameMaxSize
- used ArgumentOutRangeException instead of MarshalException for the local exception
- used LimitExceededException instead of MarshalException for the remote exception
- added minimum limit for IncomingFrameMaxSize of 1KB
- fixed OA bug where IncomingFrameMaxSize wasn't set on nameless OAs.
- minor review fixes

Please let me know how to rename the exceptions if these are still not appropriate.

One other thing: should we add Connection.PeerIncomingFrameMaxSize?

bernardnormier commented on csharp/src/Ice/SocketStream.cs:
```diff
@@ -383,6 +407,24 @@ internal virtual async ValueTask SendInitializeFrameAsync(CancellationToken canc
+            {
+                if (frame is OutgoingRequestFrame)
+                {
+                    throw new ArgumentOutOfRangeException(
```
`ArgumentOutOfrangeException` is not the appropriate exception here. From its description:

```
An ArgumentOutOfRangeException exception is thrown when a method is invoked and at least one of the arguments passed to the method is not null and contains an invalid value that is not a member of the set of values expected for the argument. The ParamName property identifies the invalid argument, and the ActualValue property, if a value is present, identifies the invalid value.```


What about:
- make LimitExceededException a new _local_ C# exception
- add a new remote exception `ServerException` which means the server-side implementation failed, but unlike UnhandledException, may not correspond to a bug.
- update Proxy.ForwardAsync to catch the local LimitExceededException and rethrow it as a `ServerException`

bentoi replied:
> Yes, we should. It's of course ice2-only. For ice1, we simply return 0 or -1, meaning N/A?

I don't have a preference for the value. 

For the `Connection.IdleTimeout` property, the setter throws `NotSupportedException("setting IdleTimeout is not supported with Slic")` for Slic connections. Should we consider throwing this exception here as well for consistency?

I'll fix the exceptions as suggested.

bernardnormier replied:
I assume `Connection.PeerIncomingFrameMaxSize` would be get-only and throwing exceptions from getters is not a recommended pattern.

bernardnormier reviewed (approved):
Looks good. See suggestion for exception names.

> One other thing: should we add Connection.PeerIncomingFrameMaxSize?

Yes, we should. It's of course ice2-only. For ice1, we simply return 0 or -1, meaning N/A?

externl reviewed (approved):
Looks good!
